src/librustc_trans/mir/block.rs

```diff
@@ -49,70 +48,130 @@ impl<'a, 'tcx> MirContext<'a, 'tcx> {
 
         debug!("trans_block({:?}={:?})", bb, data);
 
-        let funclet = match self.cleanup_kinds[bb] {
-            CleanupKind::Internal { funclet } => funclets[funclet].as_ref(),
-            _ => funclets[bb].as_ref(),
-        };
+        for statement in &data.statements {
+            bcx = self.trans_statement(bcx, statement);
+        }
+
+        self.trans_terminator(bcx, bb, data.terminator(), funclets);
+    }
+
+    fn trans_terminator(&mut self,
+                        mut bcx: Builder<'a, 'tcx>,
+                        bb: mir::BasicBlock,
+                        terminator: &mir::Terminator<'tcx>,
+                        funclets: &IndexVec<mir::BasicBlock, Option<Funclet>>)
+    {
+        debug!("trans_terminator: {:?}", terminator);
 
         // Create the cleanup bundle, if needed.
+        let tcx = bcx.tcx();
+        let span = terminator.source_info.span;
+        let funclet_bb = match self.cleanup_kinds[bb] {
```

Comment:
You added a method (`funclet_bb`) to `CleanupKind` that does this, shouldn't we just call it here? Unless I'm missing some detail.